Document Magic

engineer124 opened the pull request:
This documents the different magic variables and functions in the game, primarily from z_parameter.

A few notes:

- There are two variables to store the magicCapacity. One of them is always the true capacity, either 0, 0x30, or 0x60. The second one is the `magicCapacity` that's "drawn", which exists as a separate variable specifically for when you enter the first scene from a new load, and the magic bar grows from 0 width to the width of the current bar. I went with `magicCapacity` for the fixed value and `magicCapacityDrawn` as the drawn value. Other than that first scene load, they're always the same and are used interchangeably throughout the repo in terms of checking for magic capacity. Maybe another name could be `magicCapacityTarget` for the fixed values and `magicCapacity` for the drawn values
- I'm not sure if the different types of border changes should be encapsulated in the enums or comments, since the enum names are getting long as is
- I think the enum names for `MagicBarChange` could be improved, but I went with a more verbose name to start just to help reviewers understand how magic is working
- Some of the comments for the magic bar enums are repeated both where the enum is defined and where the switch is, I wasn't sure which of the two the comments should be. So I put them in both for now

Dragorn421 reviewed (commented):
Great job! 🪄 🎩 🐰

engineer124 commented on src/code/z_parameter.c:
```diff
@@ -2239,111 +2239,126 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
-s32 func_80087708(GlobalContext* globalCtx, s16 arg1, s16 arg2) {
-    if (!gSaveContext.magicAcquired) {
-        return 0;
+s32 Magic_ChangeBy(GlobalContext* globalCtx, s16 magicChange, s16 changeType) {
```
I was actually inspired by `Health_ChangeBy` when naming this function. I acknowledge the implementation details are slightly different, and it might be worth adding a comment above both functions emphasizing the different nature of the `XXXChange` arg1 required by each.

Maybe something like `Magic_ChangeByTarget` to emphasize that this function doesn't change magic directly, but sets a target for `Interface_UpdateMagicBar` to act upon?

engineer124 replied:
What about `Magic_RequestChange`?

Dragorn421 replied:
I guess we expect setup functions to not be able to fail (like, this function can return false if not enough magic for example), so I guess `Magic_RequestChange` is a good idea

fig02 replied:
request sounds good
I was never a fan of the `healthChange` arg name, and would probably like to see it changed for both health and magic. even just adding the word "amount" makes it much better, the amount to change by

fig02 replied:
interestingly you can also argue that Health_ChangeBy is a request cause it will fail and return false when health is already less than or equal to 0. But its less impactful than the amount of conditions that magic changing can fail under

engineer124 replied:
Could it just be 
`magicChange` -> `amount`
`changeType` -> `type`
and it's implied? The only thing is to ensure `amount` for magic is always positive since increasing or decreasing is determined by type, which is different than Health in which you use use either a positive or negative value

fig02 replied:
that could work, it could also be `changeAmount` and `changeType`  for maximum clarity 
health could also use `changeAmount`

that note about negative magic values is def important for a function level comment

fig02 replied:
id actually be fine with `amount` and `type` tbh

fig02 replied:
actually wait, `amount` could be interpreted to mean a new value to set to as opposed and amount to change by?

engineer124 replied:
Ah so you don't want it to be interpreted as a target instead of a change

fig02 commented on include/z64save.h:
```diff
@@ -4,6 +4,32 @@
+typedef enum MagicBarChange {
+    /* 00 */ MAGIC_BAR_CONSUME_NOW, // Consume Magic immediately without preview
+    /* 01 */ MAGIC_BAR_CONSUME_WAIT_NO_PREVIEW, // Sets consume target but waits to consume. No yellow magic preview to target consumption. Unused
+    /* 02 */ MAGIC_BAR_CONSUME_NOW_ALT, // Identical behaviour to MAGIC_BAR_CONSUME_NOW. Unused
+    /* 03 */ MAGIC_BAR_CONSUME_LENS, // Lens consumption
+    /* 04 */ MAGIC_BAR_CONSUME_WAIT_PREVIEW, // Sets consume target but waits to consume. Draws yellow magic to target consumption
+    /* 05 */ MAGIC_BAR_ADD // Sets a target to add magic
+} MagicBarChange;
```
would call the enum ChangeType instead of just "Change". may need to update variables to reflect that.
also normally we would have a common prefix for these values that all fit under the same umbrella, but its a bit weird with `add` being different from all the other consume types i guess

in general I think we can probably afford to make all the magic related enums just start with `MAGIC_` and omit the bar? that goes for the enum above too.
Then we can do `MAGIC_CHANGE_CONSUME_NOW` and `MAGIC_CHANGE_ADD` etc, with `MAGIC` telling you its for the magic system, and `CHANGE` telling you its for the change type, followed by the actual type name

engineer124 replied:
For your first point:
I also had `CHANGE_` in the enum earlier, but then decided to take it out because the names were getting quite long and it seems like some of the names had reduncency in them i.e.
```
typedef enum MagicBarChange {
    /* 00 */ MAGIC_BAR_CHANGE_CONSUME_NOW,
    /* 01 */ MAGIC_BAR_CHANGE_CONSUME_WAIT_NO_PREVIEW,
    /* 02 */ MAGIC_BAR_CHANGE_CONSUME_NOW_ALT,
    /* 03 */ MAGIC_BAR_CHANGE_CONSUME_LENS,
    /* 04 */ MAGIC_BAR_CHANGE_CONSUME_WAIT_PREVIEW,
    /* 05 */ MAGIC_BAR_CHANGE_ADD
} MagicBarChange;
```
The terms `CHANGE_CONSUME` and `CHANGE_ADD` feel redundant, and there's only one that's different than the others `ADD` (I'm fine with `MagicBarChangeType` though)

For your second point, I messed around a bit with `magic` vs. `magicBar` for many of these names and enums. The reason I leaned more towards `MagicBar` when I could is because `magic` felt a bit too general, like magic spell or magic amount, it's hard to articulate. I felt `MagicBar` was really clear in what it was referring to, does that make sense?

Dragorn421 commented on include/z64save.h:
```diff
@@ -4,6 +4,32 @@
+    /* 05 */ MAGIC_BAR_ADD // Sets a target to add magic
+} MagicBarChange;
+
+#define MAGIC_HALF_BAR 0x30
+#define MAGIC_FULL_BAR (2 * MAGIC_HALF_BAR)
```
I thought the same until I saw:

https://github.com/engineer124/oot/blob/764a102a863d828c3d72227a49bac6d553b60bd9/src/code/z_parameter.c#L2244
```c
gSaveContext.magicCapacity = (gSaveContext.isDoubleMagicAcquired + 1) * MAGIC_HALF_BAR;
```
, and https://github.com/engineer124/oot/blob/764a102a863d828c3d72227a49bac6d553b60bd9/src/code/z_parameter.c#L2377
```c
temp = gSaveContext.magicLevel * MAGIC_HALF_BAR; // magicCapacity
```

engineer124 commented on include/z64save.h:
```diff
@@ -4,6 +4,32 @@
+    /* 07 */ MAGIC_BAR_ACTION_LENS_CONSUME, // Magic slowly consumed by lens. 
+    /* 08 */ MAGIC_BAR_ACTION_GROW_WIDE, // Init magic on a new load, grow from a width of 0 to magicCapacity
+    /* 09 */ MAGIC_BAR_ACTION_FILL, // Add magic until full capacity is reached
+    /* 10 */ MAGIC_BAR_ACTION_ADD // Add requested magic
+} MagicBarAction;
```
I first named it "state", so I'm not opposed to the name. I decided to switch to "action" because it seemed like it was describing the magic bar actually "doing" something as oppose to "state" which sounds more like a passive description.

I would guess "state" is probably more consistent with other things in the repo, although you'd have a better sense of that

engineer124 replied:
I guess let's take a look at the spin attack as an example.

This is the line where you specify you want magic to be consumed. But this only sets the target, and previews that target with the yellow bar drawn between `magic` and `magicTarget` and it remains that way until an external trigger occurs
https://github.com/zeldaret/oot/blob/764a102a863d828c3d72227a49bac6d553b60bd9/src/overlays/actors/ovl_En_M_Thunder/z_en_m_thunder.c#L162

And that trigger is this line below. Once the magic is ready to actually be consumed, the action is set, triggering the magic to be consumed. It's almost like this variable causes the magic bar to "act" in certain ways. It felt like it was being used more than just a state, it's actually used as a trigger to get the magic bar to act.
https://github.com/zeldaret/oot/blob/764a102a863d828c3d72227a49bac6d553b60bd9/src/overlays/actors/ovl_En_M_Thunder/z_en_m_thunder.c#L196

However, I will say that if state is used like this in other places and that you can set the state to trigger certain events to occur, instead of the state being used to passively describe what's happening to a system, then I might be fine with state

fig02 replied:
action to me is just very specifically a word to describe what something more "physical" is doing, like an actor that is representing a real world thing. For a more abstract thing under the hood, state or mode seems more appropriate to me. 
I used mode a bunch in the file select system to describe the various things its doing. But tharo used state to describe the message state which i feel is pretty analogus here. providing examples from things by the same dev cause this guy like to assign state to numbers in this way pretty consistently lol

engineer124 commented:
Thoughts on consistently using the `Magic_` prefix?
`Interface_UpdateMagicBar` -> `Magic_UpdateBar`
`Interface_DrawMagicBar` -> `Magic_DrawBar`
I know the magic bar itself is part of the interface system, but it's also part of the magic system.

Also, maybe `Magic_ResetMagicBarAction` could be simplified to `Magic_ResetBar` to keep it simple?

engineer124 commented:
I've made the suggestions from the discord discussion, using the `Magic_` prefix and dropping the bar everywhere. I've also changed `ACTION` -> `STATE`, and renamed `lensMagicDepletionTimer` to just `magicConsuptionTimer`. Other discussions are still open though

Dragorn421 commented on include/z64save.h:
```diff
@@ -143,11 +169,11 @@ typedef struct {
-    /* 0x13F6 */ s16 unk_13F6; // magic related
-    /* 0x13F8 */ s16 unk_13F8; // magic related
+    /* 0x13F0 */ s16 magicState; // Determines magic meter behavior on each frame
+    /* 0x13F2 */ s16 magicStateStored; // Stores the previous magic state while magic is increasing. Allows magicState to be restored afterwards
```
bump on the suggestion for naming this `magicStateAfterFill` (or `nextMagicStateAfterFill` ?) (maybe Fill->Increase ?)

engineer124 replied:
I'm still not sure of associating this specifically with fill, since it's used with magic jars when adding instead of filling (which yes calls fill specifically to set this variable before overwritting the state)

Dragorn421 replied:
Yeah I know, I was thinking "fill" as in "anything that fills up magic even just a bit" but that indeed may be confusing with fill being already used as fully-fill

Maybe (next)`MagicStateAfter`(Increase/Add) then?

engineer124 replied:
Yeah, "fill" in magic context specifically means "increase to full capacity", while "add" in magic context specifically means "anything that fills up magic even a bit" (I comment because a different word that "fill" can be used here if you have any other suggestions).

Something like `magicStateAfterIncrease` would capture both cases

Dragorn421 commented on include/z64save.h:
```diff
@@ -4,6 +4,32 @@
+    /* 0x0 */ MAGIC_STATE_IDLE, // Regular gameplay
+    /* 0x1 */ MAGIC_STATE_CONSUME_SETUP, // Sets the speed in which magic border flashes
+    /* 0x2 */ MAGIC_STATE_CONSUME, // Consume magic until target is reached or no more magic is available
+    /* 0x3 */ MAGIC_STATE_BORDER_CHANGE_1, // Flashes border and freezes Dark Link
```
I think for the name we can abstract that it's just the border flashing and not the whole meter, to name this like `meter_flash`?

Dragorn421 commented on src/code/z_parameter.c:
```diff
@@ -2532,90 +2562,91 @@ void Interface_UpdateMagicBar(GlobalContext* globalCtx) {
         if (gSaveContext.healthCapacity > 0xA0) {
-            magicBarY = R_MAGIC_BAR_LARGE_Y;
+            magicMeterY = R_MAGIC_METER_LARGE_Y; // two rows of hearts
         } else {
-            magicBarY = R_MAGIC_BAR_SMALL_Y;
+            magicMeterY = R_MAGIC_METER_SMALL_Y; // one row of hearts
```
Not your changes but,
when I first saw these regs' name I thought it refered to half/double magic

My best suggestion is unfortunately only `R_MAGIC_METER_Y_FEW_HEARTS` and `R_MAGIC_METER_Y_MANY_HEARTS` 😅

engineer124 replied:
Yes, I see the confusion and think it should be changed. Maybe something like:
`R_MAGIC_METER_Y_POS_UPPER` (= 34)
`R_MAGIC_METER_Y_POS_LOWER` (= 42)
(Upper is a smaller number because it's coordinates from the top of the screen I'm assuming?)

Dragorn421 replied:
Yeah screen coordinates are 0,0 top left and grow towards bottom right

I don't like upper/lower because it describes the data rather than describing what the data is for. (for example github's green comment button is better described as "reply button" rather than "green box with black text")

engineer124 replied:
The UPPER is describing the Y_POS which is what this data is for? Would `R_MAGIC_METER_Y_UPPER_POS` be clearer?

Dragorn421 commented on src/code/z_parameter.c:
```diff
@@ -4072,23 +4103,26 @@ void Interface_Update(GlobalContext* globalCtx) {
         ((globalCtx->csCtx.state == CS_STATE_IDLE) || !Player_InCsMode(globalCtx))) {
-        if ((gSaveContext.magicAcquired != 0) && (gSaveContext.magicLevel == 0)) {
-            gSaveContext.magicLevel = gSaveContext.doubleMagic + 1;
-            gSaveContext.unk_13F0 = 8;
+        // Condition for updating magic
+
+        if (gSaveContext.isMagicAcquired && (gSaveContext.magicLevel == 0)) {
```
I'm not sure which conditional this comment applies to?

engineer124 commented on src/code/z_parameter.c:
```diff
@@ -2256,115 +2262,129 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
-    if ((gSaveContext.unk_13F0 != 8) && (gSaveContext.unk_13F0 != 9)) {
-        if (gSaveContext.unk_13F0 == 10) {
-            gSaveContext.unk_13F2 = gSaveContext.unk_13F0;
+void Magic_Reset(GlobalContext* globalCtx) {
```
Or what about `MAGIC_STATE_RESET` instead of `MAGIC_STATE_RESTORE_IDLE`?

Dragorn421 replied:
My first thought was `MAGIC_STATE_RESET` looks like a duplicate of `MAGIC_STATE_IDLE`, but looking at it longer yeah I think that could work :+1:

Dragorn421 reviewed (commented):
minor things besides `Magic_Reset`

Dragorn421 reviewed (approved):
Great job on this! :mage: :balloon: 

I'm still not convinced by `Magic_Reset` but I don't have a strong suggestion

fig02 commented on include/functions.h:
```diff
@@ -1040,8 +1040,8 @@ s32 Health_ChangeBy(GlobalContext* globalCtx, s16 healthChange);
 void Magic_Fill(GlobalContext* globalCtx);
-void func_800876C8(GlobalContext* globalCtx);
-s32 func_80087708(GlobalContext* globalCtx, s16 arg1, s16 arg2);
+void Magic_Reset(GlobalContext* globalCtx);
+s32 Magic_RequestChange(GlobalContext* globalCtx, s16 magicChange, s16 changeType);
```
arg names need to be updated
I also think its appropriate to change `Health_ChangeBy` to use `amount` for its arg with these set of changes

EllipticEllipsis commented on include/z64save.h:
```diff
@@ -74,14 +100,14 @@ typedef struct {
     /* 0x0038 */ u16 naviTimer;
-    /* 0x003A */ u8 magicAcquired;
+    /* 0x003A */ u8 isMagicAcquired;
     /* 0x003B */ char unk_3B[0x01];
-    /* 0x003C */ u8 doubleMagic;
+    /* 0x003C */ u8 isDoubleMagicAcquired;
```
I didn't realise this was used extensively in the repo, but now I'm wondering if we want to name variables like this interrogatively... does the `is` actually do anything here, that the plain past participle doesn't?

engineer124 replied:
I think in this specific instance, `doubleMagicAcquired` alone can be inferred to be a boolean-like value (it still uses the 0/1 in calculating `magicCapacity` so it's not a true bool), but as you mentioned, `is` is just a common and fast way to identify that something is boolean without thinking about it, at least that was my understanding of the convention

engineer124 replied:
Also as a side note, not related to this comment but something I wanted to note:
`doubleDefense` -> `isDoubleDefenseAcquired` (or without the `is`) would be consistent with `isDoubleMagicAcquired` but I refrained from changing it for now

EllipticEllipsis replied:
If we were going by the game it should be "half damage" (or possibly "enhanced defence") anyway:
```
"I'm going to enhance your \n"
COLOR(LIGHTBLUE) "defensive power" COLOR(DEFAULT) "."
[...]
"Your " COLOR(LIGHTBLUE) "defensive power" COLOR(DEFAULT) " is enhanced!"
[...]
"Your defensive power has been \n"
"enhanced! Damage inflicted by \n"
"enemies will be " COLOR(RED) "reduced by half" COLOR(DEFAULT) "."
```

fig02 replied:
imo this is a case where the common colloquial name makes more sense. no one says half damage or enhanced defense

EllipticEllipsis replied:
I had a look into this, since I was curious. I looked at 4 guidebooks (3 English, 1 the official one, 1 Japanese) and the first page of Google results for walkthroughs, including 4 GameFAQs guides. A large majority say one or more of
- half damage
- enhanced defense
- doubled health
and only 2 (of about 20) say double/doubled defense. I was pretty surprised by this myself, since the tiny speedrunning corner of the world does seem to call always call it "double defense", probably because of the alliteration.

engineer124 replied:
Just to add a different perspective, I come from the randomizer community and never heard of any of those terms, it's only ever been "double defense" for me

Dragorn421 replied:
Other related names in decomp: `FAIRY_UPGRADE_HALF_DAMAGE` (hi elliptic), `gGreatFairyDoubleDefenceCs` (hi louis)

I would be okay with normalizing all of that to "half damage", with a comment on the save context struct member like `// double defense`

EllipticEllipsis commented on src/code/z_parameter.c:
```diff
@@ -2267,115 +2273,134 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
+ * amount is the positive-valued amount to either increase or decrease magic by
+ * type is how the magic is increased or consumed
+ */
+s32 Magic_RequestChange(GlobalContext* globalCtx, s16 amount, s16 type) {
```
Could you add a note on what the return does, or at least that it's boolean?

EllipticEllipsis commented on src/code/z_parameter.c:
```diff
@@ -2151,25 +2157,25 @@ void Interface_LoadActionLabelB(GlobalContext* globalCtx, u16 action) {
 }
 
-s32 Health_ChangeBy(GlobalContext* globalCtx, s16 healthChange) {
+s32 Health_ChangeBy(GlobalContext* globalCtx, s16 amount) {
```
Here too, actually.

EllipticEllipsis commented on src/code/z_parameter.c:
```diff
@@ -2267,115 +2273,134 @@ void Inventory_ChangeAmmo(s16 item, s16 ammoChange) {
         { 150, 150, 150 },
-        { 255, 255, 150 },
-        { 255, 255, 50 },
+        { 255, 255, 150 }, // unused color
+        { 255, 255, 50 },  // unused color
```
```suggestion
        { 255, 255, 150 }, // unused
        { 255, 255, 50 },  // unused
```
Since we already know they're colours from the array name.

EllipticEllipsis commented on include/z64save.h:
```diff
@@ -143,11 +169,11 @@ typedef struct {
-    /* 0x13F8 */ s16 unk_13F8; // magic related
+    /* 0x13F0 */ s16 magicState; // determines magic meter behavior on each frame
+    /* 0x13F2 */ s16 prevMagicState; // used to resume the previous state after adding or filling magic
+    /* 0x13F4 */ s16 magicCapacityDrawn; // only differs from magicCapacity in a new save load, where magicCapacityDrawn is slowly stepped from 0 to magicCapacity
```
I think you could just say "increased" instead of "stepped" here: it's a discrete variable, yes, but it looks continuous enough in-game, it's not like you see it increase 8 pixels, stop, then do it again, or whatever.

engineer124 replied:
What about "increased every frame"? I used step because it's increased every update function over time, so I wanted to capture that in a way

EllipticEllipsis replied:
Maybe describe the visuals a bit more? It's not clear just from here that it's explicitly to show the bar itself increasing in size.
```suggestion
    /* 0x13F4 */ s16 magicCapacityDrawn; // only differs from magicCapacity in a new save load, where magicCapacityDrawn is increased from 0 to magicCapacity to show the meter size growing to its actual maximum value.
```
?

Dragorn421 replied:
"gradually increased" (a.k.a. "stepped" I guess) makes it more clear to me there's an animation going on

EllipticEllipsis commented on src/code/z_parameter.c:
```diff
@@ -2385,29 +2410,32 @@ void Interface_UpdateMagicBar(GlobalContext* globalCtx) {
+
+    switch (gSaveContext.magicState) {
+        case MAGIC_STATE_GROW_METER:
+            // Step magicCapacityDrawn to what is magicCapacity
```
As here: it's basically the same as MAGIC_STATE_FILL, but for the bar itself, right?

engineer124 replied:
In a sense. The way this block is designed is that it just moves the bar drawn to where the true capacity is. In theory, it could be moved wider or smaller (i.e. it doesn't specifically have to grow), it's just that in practice it only grows based on how this case is used

EllipticEllipsis replied:
So it could be `UPDATE_WIDTH`, even.

engineer124 replied:
`WIDTH` could apply to a lot of things so I think it'd be important to specify bar in there, but yeah I'd be fine with that. Could maybe even be `SYNC_BAR_WIDTH` in the sense that it "sync"s the `magicCapacityDrawn` to the true `magicCapacity` just as an idea

EllipticEllipsis reviewed (commented):
Nice job, this is actually comprehensible now. Couple of things.

engineer124 commented on src/overlays/actors/ovl_Boss_Ganon/z_boss_ganon.c:
```diff
@@ -1191,7 +1191,7 @@ void BossGanon_SetupTowerCutscene(BossGanon* this, GlobalContext* globalCtx) {
         this->csState = 100;
         this->unk_198 = 1;
-        gSaveContext.magic = gSaveContext.unk_13F4;
+        gSaveContext.magic = gSaveContext.magicCapacityDrawn;
```
I'd be okay with either, but I have a slight preference to the way I have it now but it’s hard to articulate why.

But I interpret it as the point of `magicCapacity` being to store the “true” value in a sense while `magicCapacityDrawn` gives flexibility on how the capacity is drawn from 0 -> `magicCapacity` or the previous `magicCapacity` -> new `magicCapacity`. Most checks when comparing against magic capacity seems to be against `magicCapacity` instead of `magicCapacityDrawn`, but either work due to them being mostly the same. But I still has a sense that `magicCapacity` is the true value and calling that `magicCapacityTarget` or `magicCapacityMax` seems a bit funny to me. 

Sure `magicCapacityDrawn` is used more to draw, but it also contains info on the magicCapacity which is what it’s used for, and is still in the name, so I don't find it misleading imo.

With the current names, you also get a meaningful distinction between the two variables instead of something like `magicCapacityMax` and `magicCapacity` which seems too similar.

There also doesn’t seemed to be a logic as to whether devs used `magicCapacity` vs `magicCapacityDrawn` (outside z_parameter) since they basically contain the same info 99.5% of the time and the devs names were awful so I’d understand if they mixed them up a little.

engineer124 replied:
To note, I followed through on my suggestion of changing `MAGIC_STATE_GROW_METER` to `MAGIC_STATE_SYNC_METER_WIDTH`, but if these variables were to change, maybe it's be worth changing it to something like `MAGIC_STATE_SYNC_CAPACITIES` or something

Roman971 replied:
I agree they are both used as the capacity in different instances, but to me it's misleading to have `magicCapacityDrawn` if it gets used in actual game logic. Personally I would interpret that name as something that is used for drawing on the hud, which means I could change its value without affecting the rest of the game, but that would be a big mistake in this case.

Regarding the names I suggested, it's true that `magicCapacity` and `magicCapacityMax`/`magicCapacityTarget` are pretty similar, but I would argue there is no real way around that since they are used for similar purposes like you mentioned, and we can't fix that on our end unfortunately.

Also, looking at the code again, `magicCapacityDrawn` is actually used more than `magicCapacity`, even when only accounting for game logic purposes. Besides debug prints, `magicCapacity` is only read in one place for case `MAGIC_STATE_FILL` in `Magic_Update`, while `magicCapacityDrawn` is used everywhere else, including when adding magic through `Magic_RequestChange`. So imo it makes more sense to treat it as the "true" capacity in almost all cases (hence `magicCapacity`). And maybe we could have the other one as `magicFillCapacity` since that's only thing it's really used for?

engineer124 replied:
Right, I see what you mean.

There's also the process of obtaining magic and double magic:
For obtaining normal magic,
`gSaveContext.magicCapacity = MAGIC_NORMAL_METER;`
Then in `Interface_Update`:
```
if (gSaveContext.isMagicAcquired && (gSaveContext.magicLevel == 0)) {
    gSaveContext.magicLevel = gSaveContext.isDoubleMagicAcquired + 1;
    gSaveContext.magicState = MAGIC_STATE_SYNC_METER_WIDTH;
```
Which is true because the `magicLevel` was 0 before, and this `magicState` is the one that steps `magicCapacityDrawn` -> `magicCapacity` (it happens when the hud is hidden so you don't actually visually see the magic bar grown).

And when obtaining double magic:
```
gSaveContext.magicCapacity = MAGIC_DOUBLE_METER;
gSaveContext.magicLevel = 0;
```
They explicitly set magicLevel to 0 so that the same code in `Interface_Update` runs i.e. `(gSaveContext.magicLevel == 0)` passes, and the state becomes `MAGIC_STATE_SYNC_METER_WIDTH`. to step `magicCapacityDrawn` -> `magicCapacity`

In both cases, `magicCapacity` is acting like a target, and `magicCapacityDrawn` does feel like the true magic capacity.


So I kinda like:
`magicCapacity` -> `magicCapacityTarget`
`magicCapacityDrawn` -> `magicCapacity`
as originally had. Then maybe:
`MAGIC_STATE_SYNC_METER_WIDTH` -> `MAGIC_STATE_SYNC_CAPACITY`

Thoughts?

engineer124 replied:
Or maybe something like:
`MAGIC_STATE_SYNC_METER_WIDTH` -> 
`MAGIC_STATE_STEP_CAPACITY`
OR
`MAGIC_STATE_UPDATE_CAPACITY`

Dragorn421 commented on src/code/z_parameter.c:
```diff
@@ -2383,29 +2411,34 @@ void Interface_UpdateMagicBar(PlayState* play) {
-                    gSaveContext.unk_13F4 += 8;
-                    if (gSaveContext.unk_13F4 > temp) {
-                        gSaveContext.unk_13F4 = temp;
+    s16 temp; // target for magicCapacity or magicBorderIndex
```
```suggestion
    s16 temp; // target for magicCapacity, or magicBorderIndex
```
as in `(target for magicCapacity) or (magicBorderIndex)`?
instead of `target for (magicCapacity or magicBorderIndex)` 😎

Dragorn421 commented on src/overlays/actors/ovl_Bg_Dy_Yoseizo/z_bg_dy_yoseizo.c:
```diff
@@ -709,16 +709,18 @@ void BgDyYoseizo_Give_Reward(BgDyYoseizo* this, PlayState* play) {
-                gSaveContext.unk_13F6 = 0x30;
+                gSaveContext.isMagicAcquired = true;
+                gSaveContext.magicFillTarget = MAGIC_NORMAL_METER;
+                // magicLevel is already 0, which triggers magicCapacity to grow
```
```suggestion
                // magicLevel is already 0, setting isMagicAcquired to true triggers magicCapacity to grow
```

Dragorn421 commented on src/overlays/gamestates/ovl_file_choose/z_file_choose.c:
```diff
@@ -1488,16 +1488,15 @@ void FileChoose_LoadGame(GameState* thisx) {
-    gSaveContext.magicLevel = gSaveContext.magic;
-
-    if (1) {}
+    // Set the fill target to be the magic from the previous save
```
```suggestion
    // Set the fill target to be the saved magic amount
```

Dragorn421 commented on src/overlays/gamestates/ovl_file_choose/z_file_choose.c:
```diff
@@ -1488,16 +1488,15 @@ void FileChoose_LoadGame(GameState* thisx) {
-    if (1) {}
+    // Set the fill target to be the magic from the previous save
+    gSaveContext.magicFillTarget = gSaveContext.magic;
+    // Set magicLevel and magic to 0 so the magic meter and magic grows from nothing to the state before the save
```
```suggestion
    // Set `magicLevel` and `magic` to 0 so `magicCapacity` then `magic` grows from nothing to respectively the full capacity and `magicFillTarget`
```
it's a lower level description idk 🤔

Dragorn421 commented on src/overlays/gamestates/ovl_select/z_select.c:
```diff
@@ -20,10 +20,9 @@ void Select_LoadGame(SelectContext* this, s32 entranceIndex) {
-        gSaveContext.unk_13F6 = gSaveContext.magic;
-        gSaveContext.magic = 0;
-        gSaveContext.unk_13F4 = 0;
-        gSaveContext.magicLevel = gSaveContext.magic;
+        gSaveContext.magicFillTarget = gSaveContext.magic;
+        gSaveContext.magicCapacity = 0;
+        gSaveContext.magicLevel = gSaveContext.magic = 0;
```
Should there be a comment like "see similar code in `FileChoose_LoadGame` for what this achieves" ?

Dragorn421 reviewed (commented):
yay figured it out for real this time
comments, mostly on comments
